File: chapter3/AngularShop/AngularShop/Models/ProductCategory.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace AngularShop.Models
{
    public class ProductCategory
    {
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public long Id { get; set; }

        [ForeignKey("Product")]
        public long ProductId { get; set; }
        [ForeignKey("Category")]
        public long CategoryId { get; set; }

        /// <summary>
        /// 상품이 여러 카테고리에 속할 경우, 속한 카테고리의 우선순위
        /// </summary>
        [Required]
        public int SortOrder { get; set; }

        #region 네비게이션 속성

        public virtual Product Product { get; set; }

        public virtual Category Category { get; set; }

        #endregion
    }
}

File: chapter4/AngularShop/AngularShopBoot/src/main/kotlin/com/angularShop/angularShop/models/ProductCategory.kt
package com.angularShop.angularShop.models

import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.ObjectIdGenerators
import javax.persistence.*

@Entity
@Table(name = "ProductCategories")
//중복해서 생성되는 걸 막아준다. 그런데 중복해서 나와야 하는 항목에는 사용하면 안됨.
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator::class, property = "id")
data class ProductCategory(
        @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
        var id: Long? = null,

        var sortOrder: Int,

        //현재 테이블의 productId 속성을 외래키로 사용
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name="productId")
        var product: Product,

        //현재 테이블의 categoryId 속성을 외래키로 사용
        @ManyToOne(fetch = FetchType.LAZY)
        @JoinColumn(name="categoryId")
        var category: Category
)